src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1025,46 +929,34 @@ impl Store {
     &self,
     destination: PathBuf,
     digest: Digest,
-  ) -> BoxFuture<'static, Result<DirectoryMaterializeMetadata, String>> {
-    let root = Arc::new(Mutex::new(None));
+  ) -> BoxFuture<'static, Result<(), String>> {
     self
-      .materialize_directory_helper(
-        destination,
-        RootOrParentMetadataBuilder::Root(root.clone()),
-        digest,
-      )
-      .and_then(move |()| {
-        future::ready(Ok(
-          Arc::try_unwrap(root).unwrap().into_inner().unwrap().build(),
-        ))
-      })
+      .materialize_directory_helper(destination, true, digest)
       .boxed()
   }
 
   fn materialize_directory_helper(
     &self,
     destination: PathBuf,
-    root_or_parent_metadata: RootOrParentMetadataBuilder,
+    is_root: bool,
     digest: Digest,
   ) -> BoxFuture<'static, Result<(), String>> {
     let store = self.clone();
     async move {
-      let directory_creation =
-        if let RootOrParentMetadataBuilder::Root(..) = root_or_parent_metadata {
-          let destination = destination.clone();
-          store
-            .local
-            .executor()
-            .spawn_blocking(move || fs::safe_create_dir_all(&destination))
-            .await
-        } else {
-          let destination = destination.clone();
-          store
-            .local
-            .executor()
-            .spawn_blocking(move || fs::safe_create_dir(&destination))
-            .await
-        };
+      let destination2 = destination.clone();
+      let directory_creation = if is_root {
+        store
+          .local
+          .executor()
+          .spawn_blocking(move || fs::safe_create_dir_all(&destination2))
```

Comment:
Possible simplification: If we lift the directory creation to be performed by the caller, rather than the callee, the `is_root` param can disappear completely - `materialize_directory` could `create_dir_all` the root destination, and then we could do the `create_dir` when we discover a child `DirectoryNode` rather than when we're materializing the `DirectoryNode`'s `Digest` (though possibly that makes the new added parallelism harder to add :))

Reply:
Yea: it makes the parallelism a bit easier to keep this in the body I think.